kv/dist_sender.go

```diff
@@ -531,11 +553,19 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
 func (ds *DistSender) sendAttempt(trace *tracer.Trace, args proto.Request, desc *proto.RangeDescriptor) (proto.Response, error) {
 	defer trace.Epoch("sending RPC")()
 	// Truncate the request to our current range, making sure not to
-	// touch it unless we have to (it is illegal to send EndKey on
-	// commands which do not operate on ranges).
-	if endKey := args.Header().EndKey; endKey != nil && !endKey.Less(desc.EndKey) {
-		defer func(k proto.Key) { args.Header().EndKey = k }(endKey)
-		args.Header().EndKey = desc.EndKey
+	// touch it unless we have to
+	if _, ok := args.(*proto.ReverseScanRequest); ok {
+		// If the reverse scan request is cross multiple ranges.
+		if key := args.Header().Key; key != nil && key.Less(desc.StartKey) {
+			defer func(k proto.Key) { args.Header().Key = k }(key)
+			args.Header().Key = desc.StartKey
+		}
+	} else {
+		// It is illegal to send EndKey on commands which do not operate on ranges.
+		if endKey := args.Header().EndKey; endKey != nil && !endKey.Less(desc.EndKey) {
+			defer func(k proto.Key) { args.Header().EndKey = k }(endKey)
+			args.Header().EndKey = desc.EndKey
+		}
```

Comment:
ok, I think the code is here because doing it outside of `sendAttempt` (where it belongs semantically) can't `defer()` as elegantly. Can you still pull this up? I'm commenting at the call site below.